packages/vscode-extension/src/project/project.ts

```diff
@@ -415,6 +457,32 @@ export class Project implements Disposable, ProjectInterface, DeviceSessionsMana
   }
 
   public dispatchTouches(touches: Array<TouchPoint>, type: "Up" | "Move" | "Down") {
+    const rotation = this.projectState.rotation;
+
+    // transfrom touch coordinates to account for different device orientations before
+    // sending them for dispatch to the device session.
+
+    touches.forEach((touch, i, array) => {
+      const { xRatio: x, yRatio: y } = touch;
+      switch (rotation) {
+        // 90° anticlockwise map (x,y) to (1-y, x)
+        case DeviceRotationType.LandscapeLeft:
+          array[i] = { xRatio: 1 - y, yRatio: x };
+          break;
+        case DeviceRotationType.LandscapeRight:
+          // 90° clockwise map (x,y) to (y, 1-x)
+          array[i] = { xRatio: y, yRatio: 1 - x };
+          break;
+        case DeviceRotationType.PortraitUpsideDown:
+          // 180° map (x,y) to (1-x, 1-y)
+          array[i] = { xRatio: 1 - x, yRatio: 1 - y };
+          break;
+        default:
+          // Portrait mode: no transformation needed
+          break;
+      }
+    });
+
```

Comment:
I don't think this logic belongs here.
Project is mostly a facade that the frontend can call, and (currently) handles some workspace-specific stuff (like watching the workspace configuration).
This is more about translating between the representation used by the JS part of Radon and the representation that the sim-server requires, which I think _should_ happen in `DeviceBase` (since that's the module that directly talks to it).
I'd move it down to `DeviceBase`.